subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/DefaultFileCollectionDependency.java

```diff
@@ -0,0 +1,140 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.api.internal.artifacts.dependencies;
+
+import org.gradle.api.artifacts.Dependency;
+import org.gradle.api.artifacts.FileCollectionDependency;
+import org.gradle.api.artifacts.component.ComponentIdentifier;
+import org.gradle.api.file.FileCollection;
+import org.gradle.api.internal.file.FileCollectionInternal;
+import org.gradle.api.tasks.TaskDependency;
+import org.gradle.internal.deprecation.DeprecationLogger;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.util.Set;
+
+public class DefaultFileCollectionDependency extends AbstractDependency implements SelfResolvingDependencyInternal, FileCollectionDependency {
+    private final ComponentIdentifier targetComponentId;
+    private final FileCollectionInternal source;
+
+    public DefaultFileCollectionDependency(FileCollectionInternal source) {
+        this.targetComponentId = null;
+        this.source = source;
+    }
+
+    public DefaultFileCollectionDependency(ComponentIdentifier targetComponentId, FileCollectionInternal source) {
+        this.targetComponentId = targetComponentId;
+        this.source = source;
+    }
+
+    @Override
+    public boolean contentEquals(Dependency dependency) {
+        if (!(dependency instanceof DefaultFileCollectionDependency)) {
+            return false;
+        }
+        DefaultFileCollectionDependency selfResolvingDependency = (DefaultFileCollectionDependency) dependency;
+        return source.equals(selfResolvingDependency.source);
+    }
+
+    @Override
+    public DefaultFileCollectionDependency copy() {
+        return new DefaultFileCollectionDependency(targetComponentId, source);
+    }
+
+    @Nullable
+    @Override
+    public ComponentIdentifier getTargetComponentId() {
+        return targetComponentId;
+    }
+
+    @Override
+    public String getGroup() {
+        return null;
+    }
+
+    @Override
+    public String getName() {
+        return "unspecified";
+    }
+
+    @Override
+    public String getVersion() {
+        return null;
+    }
+
+    @Override
+    @Deprecated
+    @SuppressWarnings("deprecation")
+    public void resolve(org.gradle.api.internal.artifacts.CachingDependencyResolveContext context) {
+        context.add(source);
+    }
+
+    @Override
+    @Deprecated
+    public Set<File> resolve() {
+        if (emitDeprecation()) {
+            DeprecationLogger.deprecate("Directly resolving a file collection dependency's files")
+                .withAdvice("Add the dependency to a resolvable configuration and resolve the configuration.")
```

Comment:
🤔 Is there any way to make it clearer which dependency this is talking about?

Reply:
For sure the stacktrace.

Would the targetComponentId help at all?  That seems like it would always be present except in tests, right?

Reply:
I think it is often not present. That first constructor gets called by the instantiator, so its uses do not appear in the IDE

Reply:
We can print the target component ID but in practice I think the stack trace will guide the user to a solution in almost all cases.